3D diagnostic tendency bug fixes for operational physics

SamuelTrahanNOAA opened the pull request:
Bug fixes to ldiag3d and addition of a qdiag3d flag to turn on tracer tendencies.

SamuelTrahanNOAA committed:
add qdiag3d support

SamuelTrahanNOAA committed:
several missing changes for qdiag3d support

SamuelTrahanNOAA committed:
Further bug fixes to tendency accumulation

DomHeinzeller commented on physics/GFS_GWD_generic.F90:
```diff
@@ -91,10 +92,13 @@ subroutine GFS_GWD_generic_pre_run(                               &
       if (lssav) then
-        if (ldiag3d) then
+        if (ldiag3d .and. gwd_generic_tend) then
+          write(0,*) 'gwd_generic_tend'
```
We should probably remove this write statement

DomHeinzeller commented on physics/GFS_GWD_generic.F90:
```diff
@@ -150,7 +154,8 @@ subroutine GFS_GWD_generic_post_run(lssav, ldiag3d, dtf, dusfcg, dvsfcg, dudt, d
 
-        if (ldiag3d) then
+        if (ldiag3d .and. gwd_generic_tend) then
+          write(0,*) 'gwd_generic_tend'
```
Dito

DomHeinzeller commented on physics/GFS_GWD_generic.meta:
```diff
@@ -118,6 +118,20 @@
+  units = m s-2
+  dimensions = (horizontal_dimension,vertical_dimension)
+  type = real
+  kind = kind_phys
```
I am missing the intent and optional flags for dudt and dvdt

DomHeinzeller commented on physics/GFS_GWD_generic.meta:
```diff
@@ -127,6 +141,20 @@
+  units = m s-1
+  dimensions = (horizontal_dimension,vertical_dimension)
+  type = real
+  kind = kind_phys
```
As above, but for du3dt and dv3dt

DomHeinzeller commented on physics/GFS_GWD_generic.meta:
```diff
@@ -145,6 +173,12 @@
+  long_name = true if GFS_GWD_generic should calculate tendencies
+  units = flag
+  dimensions = ()
+  type = logical
```
dito

DomHeinzeller commented on physics/GFS_GWD_generic.meta:
```diff
@@ -287,6 +321,12 @@
+  long_name = true if GFS_GWD_generic should calculate tendencies
+  units = flag
+  dimensions = ()
+  type = logical
```
dito

DomHeinzeller commented on physics/GFS_MP_generic.F90:
```diff
@@ -42,12 +42,14 @@ subroutine GFS_MP_generic_pre_run(im, levs, ldiag3d, do_aw, ntcw, nncl, ntrac, g
         do k=1,levs
           do i=1,im
             save_t(i,k) = gt0(i,k)
-            save_q(1:im,:,1) = gq0(1:im,:,1)
```
this assignment of the entire column inside the k loop didn't make any sense, thanks for fixing it

SamuelTrahanNOAA replied:
There is an error here.  The save_qv variable should be used instead.

DomHeinzeller commented on physics/cires_ugwp.meta:
```diff
@@ -850,6 +850,60 @@
+  units = m s-1
+  dimensions = (horizontal_dimension,vertical_dimension)
+  type = real
+  kind = kind_phys
```
Please add intent and optional attributes for all new variables.

DomHeinzeller commented on physics/gwdc.f:
```diff
@@ -1498,13 +1498,14 @@ subroutine gwdc_post_run(                                         &
       endif   ! end if_lssav
 
+      if (ldiag3d) then
+         write(0,*) 'update gwdc tend'
```
Please remove the write statement

DomHeinzeller reviewed (changes requested):
This is great, thanks very much. I left a few notes on code to clean up. I would also like to have @grantfirl take a look.

SamuelTrahanNOAA commented:
The gwd_generic_tend functionality is separate from the ldiag3d/qdiag3d
tendencies.  It looks like it was a half-implemented attempt at splitting
the gravity wave drag tendencies out to more terms.  I'm not sure it
works.  The two ldiag3d gravity wave drag tendencies from the days of yore
do work.

On Mon, 6 Jan 2020 at 09:36, DomHeinzeller <notifications@github.com> wrote:

> *@DomHeinzeller* requested changes on this pull request.
>
> This is great, thanks very much. I left a few notes on code to clean up. I
> would also like to have @grantfirl <https://github.com/grantfirl> take a
> look.
> ------------------------------
>
> In physics/GFS_GWD_generic.F90
> <https://github.com/NOAA-GSD/ccpp-physics/pull/5#discussion_r363315584>:
>
> > @@ -91,10 +92,13 @@ subroutine GFS_GWD_generic_pre_run(                               &
>        endif   ! end if_nmtvr
>
>        if (lssav) then
> -        if (ldiag3d) then
> +        if (ldiag3d .and. gwd_generic_tend) then
> +          write(0,*) 'gwd_generic_tend'
>
> We should probably remove this write statement
> ------------------------------
>
> In physics/GFS_GWD_generic.F90
> <https://github.com/NOAA-GSD/ccpp-physics/pull/5#discussion_r363315688>:
>
> > @@ -150,7 +154,8 @@ subroutine GFS_GWD_generic_post_run(lssav, ldiag3d, dtf, dusfcg, dvsfcg, dudt, d
>          dugwd(:) = dugwd(:) + dusfcg(:)*dtf
>          dvgwd(:) = dvgwd(:) + dvsfcg(:)*dtf
>
> -        if (ldiag3d) then
> +        if (ldiag3d .and. gwd_generic_tend) then
> +          write(0,*) 'gwd_generic_tend'
>
> Dito
> ------------------------------
>
> In physics/GFS_GWD_generic.meta
> <https://github.com/NOAA-GSD/ccpp-physics/pull/5#discussion_r363316005>:
>
> > @@ -118,6 +118,20 @@
>    type = logical
>    intent = in
>    optional = F
> +[dudt]
> +  standard_name = tendency_of_x_wind_due_to_model_physics
> +  long_name = zonal wind tendency due to model physics
> +  units = m s-2
> +  dimensions = (horizontal_dimension,vertical_dimension)
> +  type = real
> +  kind = kind_phys
>
> I am missing the intent and optional flags for dudt and dvdt
> ------------------------------
>
> In physics/GFS_GWD_generic.meta
> <https://github.com/NOAA-GSD/ccpp-physics/pull/5#discussion_r363316124>:
>
> > @@ -127,6 +141,20 @@
>    kind = kind_phys
>    intent = in
>    optional = F
> +[du3dt]
> +  standard_name = cumulative_change_in_x_wind_due_to_orographic_gravity_wave_drag
> +  long_name = cumulative change in x wind due to orographic gravity wave drag
> +  units = m s-1
> +  dimensions = (horizontal_dimension,vertical_dimension)
> +  type = real
> +  kind = kind_phys
>
> As above, but for du3dt and dv3dt
> ------------------------------
>
> In physics/GFS_GWD_generic.meta
> <https://github.com/NOAA-GSD/ccpp-physics/pull/5#discussion_r363316166>:
>
> > @@ -145,6 +173,12 @@
>    kind = kind_phys
>    intent = in
>    optional = F
> +[gwd_generic_tend]
> +  standard_name = true_if_GFS_GWD_generic_should_calculate_tendencies
> +  long_name = true if GFS_GWD_generic should calculate tendencies
> +  units = flag
> +  dimensions = ()
> +  type = logical
>
> dito
> ------------------------------
>
> In physics/GFS_GWD_generic.meta
> <https://github.com/NOAA-GSD/ccpp-physics/pull/5#discussion_r363316220>:
>
> > @@ -287,6 +321,12 @@
>    kind = kind_phys
>    intent = inout
>    optional = F
> +[gwd_generic_tend]
> +  standard_name = true_if_GFS_GWD_generic_should_calculate_tendencies
> +  long_name = true if GFS_GWD_generic should calculate tendencies
> +  units = flag
> +  dimensions = ()
> +  type = logical
>
> dito
> ------------------------------
>
> In physics/GFS_MP_generic.F90
> <https://github.com/NOAA-GSD/ccpp-physics/pull/5#discussion_r363316632>:
>
> > @@ -42,12 +42,14 @@ subroutine GFS_MP_generic_pre_run(im, levs, ldiag3d, do_aw, ntcw, nncl, ntrac, g
>          do k=1,levs
>            do i=1,im
>              save_t(i,k) = gt0(i,k)
> -            save_q(1:im,:,1) = gq0(1:im,:,1)
>
> this assignment of the entire column inside the k loop didn't make any
> sense, thanks for fixing it
> ------------------------------
>
> In physics/cires_ugwp.meta
> <https://github.com/NOAA-GSD/ccpp-physics/pull/5#discussion_r363317927>:
>
> > @@ -850,6 +850,60 @@
>    type = integer
>    intent = in
>    optional = F
> +[ldu3dt_ogw]
> +  standard_name = cumulative_change_in_x_wind_due_to_orographic_gravity_wave_drag
> +  long_name = cumulative change in x wind due to orographic gravity wave drag
> +  units = m s-1
> +  dimensions = (horizontal_dimension,vertical_dimension)
> +  type = real
> +  kind = kind_phys
>
> Please add intent and optional attributes for all new variables.
> ------------------------------
>
> In physics/gwdc.f
> <https://github.com/NOAA-GSD/ccpp-physics/pull/5#discussion_r363318037>:
>
> >        endif   ! end if_lssav
>
> +      if (ldiag3d) then
> +         write(0,*) 'update gwdc tend'
>
> Please remove the write statement
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-GSD/ccpp-physics/pull/5?email_source=notifications&email_token=AJMW4SKJZJ6J27ALSHA7SATQ4M6XNA5CNFSM4KCPQQJKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCQX3WRQ#pullrequestreview-338672454>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJMW4SOIVSWCQDWBSI7XOWDQ4M6XNANCNFSM4KCPQQJA>
> .
>

SamuelTrahanNOAA commented:
(Prior, incorrect, comment was deleted.)

The gwd_generic_tend is a flag I added to deal with inconsistencies in the
logic of GWD tendencies.  There is a second set of GWD tendencies that I
think Valary Yudin added, and I'm not sure those work.  I've left them
alone as best I can and got the two old GWD tendencies working.

On Mon, 6 Jan 2020 at 10:16, Samuel Trahan - NOAA Affiliate <
samuel.trahan@noaa.gov> wrote:

> The gwd_generic_tend functionality is separate from the ldiag3d/qdiag3d
> tendencies.  It looks like it was a half-implemented attempt at splitting
> the gravity wave drag tendencies out to more terms.  I'm not sure it
> works.  The two ldiag3d gravity wave drag tendencies from the days of yore
> do work.
>
> On Mon, 6 Jan 2020 at 09:36, DomHeinzeller <notifications@github.com>
> wrote:
>
>> *@DomHeinzeller* requested changes on this pull request.
>>
>> This is great, thanks very much. I left a few notes on code to clean up.
>> I would also like to have @grantfirl <https://github.com/grantfirl> take
>> a look.
>> ------------------------------
>>
>> In physics/GFS_GWD_generic.F90
>> <https://github.com/NOAA-GSD/ccpp-physics/pull/5#discussion_r363315584>:
>>
>> > @@ -91,10 +92,13 @@ subroutine GFS_GWD_generic_pre_run(                               &
>>        endif   ! end if_nmtvr
>>
>>        if (lssav) then
>> -        if (ldiag3d) then
>> +        if (ldiag3d .and. gwd_generic_tend) then
>> +          write(0,*) 'gwd_generic_tend'
>>
>> We should probably remove this write statement
>> ------------------------------
>>
>> In physics/GFS_GWD_generic.F90
>> <https://github.com/NOAA-GSD/ccpp-physics/pull/5#discussion_r363315688>:
>>
>> > @@ -150,7 +154,8 @@ subroutine GFS_GWD_generic_post_run(lssav, ldiag3d, dtf, dusfcg, dvsfcg, dudt, d
>>          dugwd(:) = dugwd(:) + dusfcg(:)*dtf
>>          dvgwd(:) = dvgwd(:) + dvsfcg(:)*dtf
>>
>> -        if (ldiag3d) then
>> +        if (ldiag3d .and. gwd_generic_tend) then
>> +          write(0,*) 'gwd_generic_tend'
>>
>> Dito
>> ------------------------------
>>
>> In physics/GFS_GWD_generic.meta
>> <https://github.com/NOAA-GSD/ccpp-physics/pull/5#discussion_r363316005>:
>>
>> > @@ -118,6 +118,20 @@
>>    type = logical
>>    intent = in
>>    optional = F
>> +[dudt]
>> +  standard_name = tendency_of_x_wind_due_to_model_physics
>> +  long_name = zonal wind tendency due to model physics
>> +  units = m s-2
>> +  dimensions = (horizontal_dimension,vertical_dimension)
>> +  type = real
>> +  kind = kind_phys
>>
>> I am missing the intent and optional flags for dudt and dvdt
>> ------------------------------
>>
>> In physics/GFS_GWD_generic.meta
>> <https://github.com/NOAA-GSD/ccpp-physics/pull/5#discussion_r363316124>:
>>
>> > @@ -127,6 +141,20 @@
>>    kind = kind_phys
>>    intent = in
>>    optional = F
>> +[du3dt]
>> +  standard_name = cumulative_change_in_x_wind_due_to_orographic_gravity_wave_drag
>> +  long_name = cumulative change in x wind due to orographic gravity wave drag
>> +  units = m s-1
>> +  dimensions = (horizontal_dimension,vertical_dimension)
>> +  type = real
>> +  kind = kind_phys
>>
>> As above, but for du3dt and dv3dt
>> ------------------------------
>>
>> In physics/GFS_GWD_generic.meta
>> <https://github.com/NOAA-GSD/ccpp-physics/pull/5#discussion_r363316166>:
>>
>> > @@ -145,6 +173,12 @@
>>    kind = kind_phys
>>    intent = in
>>    optional = F
>> +[gwd_generic_tend]
>> +  standard_name = true_if_GFS_GWD_generic_should_calculate_tendencies
>> +  long_name = true if GFS_GWD_generic should calculate tendencies
>> +  units = flag
>> +  dimensions = ()
>> +  type = logical
>>
>> dito
>> ------------------------------
>>
>> In physics/GFS_GWD_generic.meta
>> <https://github.com/NOAA-GSD/ccpp-physics/pull/5#discussion_r363316220>:
>>
>> > @@ -287,6 +321,12 @@
>>    kind = kind_phys
>>    intent = inout
>>    optional = F
>> +[gwd_generic_tend]
>> +  standard_name = true_if_GFS_GWD_generic_should_calculate_tendencies
>> +  long_name = true if GFS_GWD_generic should calculate tendencies
>> +  units = flag
>> +  dimensions = ()
>> +  type = logical
>>
>> dito
>> ------------------------------
>>
>> In physics/GFS_MP_generic.F90
>> <https://github.com/NOAA-GSD/ccpp-physics/pull/5#discussion_r363316632>:
>>
>> > @@ -42,12 +42,14 @@ subroutine GFS_MP_generic_pre_run(im, levs, ldiag3d, do_aw, ntcw, nncl, ntrac, g
>>          do k=1,levs
>>            do i=1,im
>>              save_t(i,k) = gt0(i,k)
>> -            save_q(1:im,:,1) = gq0(1:im,:,1)
>>
>> this assignment of the entire column inside the k loop didn't make any
>> sense, thanks for fixing it
>> ------------------------------
>>
>> In physics/cires_ugwp.meta
>> <https://github.com/NOAA-GSD/ccpp-physics/pull/5#discussion_r363317927>:
>>
>> > @@ -850,6 +850,60 @@
>>    type = integer
>>    intent = in
>>    optional = F
>> +[ldu3dt_ogw]
>> +  standard_name = cumulative_change_in_x_wind_due_to_orographic_gravity_wave_drag
>> +  long_name = cumulative change in x wind due to orographic gravity wave drag
>> +  units = m s-1
>> +  dimensions = (horizontal_dimension,vertical_dimension)
>> +  type = real
>> +  kind = kind_phys
>>
>> Please add intent and optional attributes for all new variables.
>> ------------------------------
>>
>> In physics/gwdc.f
>> <https://github.com/NOAA-GSD/ccpp-physics/pull/5#discussion_r363318037>:
>>
>> >        endif   ! end if_lssav
>>
>> +      if (ldiag3d) then
>> +         write(0,*) 'update gwdc tend'
>>
>> Please remove the write statement
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/NOAA-GSD/ccpp-physics/pull/5?email_source=notifications&email_token=AJMW4SKJZJ6J27ALSHA7SATQ4M6XNA5CNFSM4KCPQQJKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCQX3WRQ#pullrequestreview-338672454>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AJMW4SOIVSWCQDWBSI7XOWDQ4M6XNANCNFSM4KCPQQJA>
>> .
>>
>

SamuelTrahanNOAA committed:
fix several bugs mentioned in code review

SamuelTrahanNOAA commented:
I think all issues are now resolved.

I'm still not sure what to do about the duplicated GWD and PBL diagnostics variables, but I suspect that problem is much more widespread than just Valery's additions.

SamuelTrahanNOAA commented:
I just fixed additional bugs.  The code that was adding the PBL tendencies subtracted those same tendencies from the OGW tendency variables.  I had fixed that in one place, but not another.  The recent batch of CCPP updates added many new bugs to the diagnostic tendencies.  This time I think I've fixed the ones in the operational physics.

SamuelTrahanNOAA committed:
Move PBL tendencies into the PBL run subroutine.

SamuelTrahanNOAA committed:
add a missing intent(in) to physics/moninedmf.meta

DomHeinzeller commented:
@SamuelTrahanNOAA just to confirm - is this PR now ok from your side?

SamuelTrahanNOAA commented:
Yes.

On Wed, 22 Jan 2020 at 16:25, DomHeinzeller <notifications@github.com>
wrote:

> @SamuelTrahanNOAA <https://github.com/SamuelTrahanNOAA> just to confirm -
> is this PR now ok from your side?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-GSD/ccpp-physics/pull/5?email_source=notifications&email_token=AJMW4SPMP7SSADPIUX3R4STQ7C2WNA5CNFSM4KCPQQJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJVFE2Y#issuecomment-577393259>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJMW4SOVUPRMVNAY72RWHQLQ7C2WNANCNFSM4KCPQQJA>
> .
>

DomHeinzeller commented on physics/cires_ugwp.F90:
```diff
@@ -172,6 +174,12 @@ subroutine cires_ugwp_run(do_ugwp, me,  master, im,  levs, ntrac, dtp, kdt, lonr
     real(kind=kind_phys),    intent(out), dimension(im, levs):: dudt_mtb, dudt_ogw, dudt_tms
+
+    ! These arrays are only allocated if ldiag=.true.
+    real(kind=kind_phys),    intent(inout), dimension(im, levs) :: ldu3dt_ogw, ldv3dt_ogw, ldt3dt_ogw
```
If these arrays are only allocated if ldiag3d (not ldiag) is true, then they should be assumed-size arrays (:,:) instead of having explicit dimensions.

DomHeinzeller commented on physics/GFS_GWD_generic.meta:
```diff
@@ -145,6 +177,13 @@
   intent = in
   optional = F
+[gwd_generic_tend]
+  standard_name = true_if_GFS_GWD_generic_should_calculate_tendencies
```
I believe the standard name should be something like flag_for_... @grantfirl once again your opinion would be appreciated.

grantfirl replied:
Agree: flag_for_generic_gwd_tendencies is better, although "generic" is a bit vague in this context. You mean generic as opposed to tendencies that Valery put in for the CIRES UGWP scheme?

SamuelTrahanNOAA replied:
GFS_GWD_generic is the name of the scheme that calculates the tendencies if that flag is true

DomHeinzeller reviewed (changes requested):
A few minor comments. I would also get @grantfirl's opinion on moving the tendency calculation inside physics schemes, in particular moninedmf.f (Grant has been working much more on/with the various PBL schemes than I have).

SamuelTrahanNOAA commented:
The array dimension and flag name changes need to be made.  I'll work on
that.

Moving the tendency calculations into the physics schemes is necessary
since the generic interstitial code doesn't know which tendencies should
come out of a scheme.  That information is scheme-specific.

On Wed, 22 Jan 2020 at 14:38, DomHeinzeller <notifications@github.com>
wrote:

> *@DomHeinzeller* requested changes on this pull request.
>
> A few minor comments. I would also get @grantfirl
> <https://github.com/grantfirl>'s opinion on moving the tendency
> calculation inside physics schemes, in particular moninedmf.f (Grant has
> been working much more on/with the various PBL schemes than I have).
> ------------------------------
>
> In physics/cires_ugwp.F90
> <https://github.com/NOAA-GSD/ccpp-physics/pull/5#discussion_r369814849>:
>
> > @@ -172,6 +174,12 @@ subroutine cires_ugwp_run(do_ugwp, me,  master, im,  levs, ntrac, dtp, kdt, lonr
>      real(kind=kind_phys),    intent(out), dimension(im)      :: tau_mtb, tau_ogw, tau_tofd, tau_ngw
>      real(kind=kind_phys),    intent(out), dimension(im, levs):: gw_dudt, gw_dvdt, gw_dtdt, gw_kdis
>      real(kind=kind_phys),    intent(out), dimension(im, levs):: dudt_mtb, dudt_ogw, dudt_tms
> +
> +    ! These arrays are only allocated if ldiag=.true.
> +    real(kind=kind_phys),    intent(inout), dimension(im, levs) :: ldu3dt_ogw, ldv3dt_ogw, ldt3dt_ogw
>
> If these arrays are only allocated if ldiag3d (not ldiag) is true, then
> they should be assumed-size arrays (:,:) instead of having explicit
> dimensions.
> ------------------------------
>
> In physics/GFS_GWD_generic.meta
> <https://github.com/NOAA-GSD/ccpp-physics/pull/5#discussion_r369816376>:
>
> > @@ -145,6 +177,13 @@
>    kind = kind_phys
>    intent = in
>    optional = F
> +[gwd_generic_tend]
> +  standard_name = true_if_GFS_GWD_generic_should_calculate_tendencies
>
> I believe the standard name should be something like flag_for_...
> @grantfirl <https://github.com/grantfirl> once again your opinion would
> be appreciated.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-GSD/ccpp-physics/pull/5?email_source=notifications&email_token=AJMW4SJ5WCA4RYUGMHVMSD3Q7C4G3A5CNFSM4KCPQQJKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCSWYWUY#pullrequestreview-346917715>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJMW4SM3K7IFPEX6LMZ36PDQ7C4G3ANCNFSM4KCPQQJA>
> .
>

grantfirl commented on physics/GFS_DCNV_generic.meta:
```diff
@@ -25,6 +25,14 @@
   intent = in
   optional = F
+[qdiag3d]
+  standard_name = flag_tracer_diagnostics_3D
```
Isn't it just specific humidity tendencies instead of all tracers?

SamuelTrahanNOAA replied:
No.  It includes ozone tendencies.

grantfirl commented on physics/GFS_MP_generic.F90:
```diff
@@ -42,12 +42,24 @@ subroutine GFS_MP_generic_pre_run(im, levs, ldiag3d, do_aw, ntcw, nncl, ntrac, g
+              enddo
+           enddo
+        endif
+        if(do_aw) then
```
I'm curious why this special treatment for do_aw...

SamuelTrahanNOAA replied:
I don't know what do_aw is for, so I left its implementation unmodified as much as possible.

grantfirl commented on physics/GFS_MP_generic.F90:
```diff
@@ -42,12 +42,24 @@ subroutine GFS_MP_generic_pre_run(im, levs, ldiag3d, do_aw, ntcw, nncl, ntrac, g
+                 ! Here, gq0(...,1) is used instead of gq0_water_vapor
+                 ! to be consistent with the GFS_MP_generic_post_run
+                 ! code.
+                 save_qv(i,k) = gq0(i,k,1)
```
All things equal, I'd prefer to rely on passing in gq0_water_vapor instead of the whole array and hard-coding the water vapor index. I see that the whole array is used below for do_aw, in which case, OK, no need to pass in the data twice.

climbfuji replied:
Instead of `gq0(i,k,1)` we should be using `gq0(i,k,ntqv)` in this case.

grantfirl replied:
Thanks @climbfuji. I couldn't remember if there was a named index for water vapor or not. Since there is, as long as the whole tracer array is needed for do_aw, using the named index would definitely be the preferred solution.

grantfirl commented on physics/GFS_MP_generic.F90:
```diff
@@ -256,9 +269,15 @@ subroutine GFS_MP_generic_post_run(im, ix, levs, kdt, nrcm, ncld, nncl, ntcw, nt
+          if (qdiag3d) then
+             do k=1,levs
+                do i=1,im
+                   dq3dt(i,k) = dq3dt(i,k) + (gq0(i,k,1)-save_qv(i,k)) * frain
```
similar comment to above. I'd prefer to see just the water vapor slice passed in.

grantfirl commented on physics/moninedmf.f:
```diff
@@ -1037,6 +1041,17 @@ subroutine hedmf_run (ix,im,km,ntrac,ntcw,dv,du,tau,rtg,          &
             rtg(i,k,1) = rtg(i,k,1)+qtend
             dtsfc(i)   = dtsfc(i)+cont*del(i,k)*ttend
             dqsfc(i)   = dqsfc(i)+conq*del(i,k)*qtend
+            if(lssav .and. ldiag3d) then
```
I'm a little confused why tendencies specific to moninedmf are needed. Shouldn't the ones calculated in GFS_PBL_generic suffice? If not, would you want to add ".and. .not. pbl_generic_tend"?

SamuelTrahanNOAA replied:
The pbl_generic_tend should be in there.  I will fix that.

grantfirl commented:
I'm curious what the strategy for tendencies in general is. Process-split schemes like surface layer, PBL, and GWD must output a tendency while time-split schemes like the others don't necessarily have to except for diagnostic purposes, although they are certainly good diagnostics to have. My opinion (not that anyone asked for it) would be to basically treat all CCPP schemes like process-split schemes and to calculate tendencies by default automatically through the CCPP framework and to not allow schemes to modify the state internally. For time-split schemes, those tendencies could then simply be applied automatically after they're called too.

With the exception of CIRES UGWP, we've been putting all diagnostic tendencies in the "scheme-generic" interstitials because presumably we'd like to have tendencies calculated in the same way for any scheme of a given type. IMO, relying on all schemes to calculate their own tendencies internally means a lot of repeated code and potential confusion...

SamuelTrahanNOAA commented:
No.  Ozone is in the dq3dt array.

On Thu, 23 Jan 2020 at 15:09, grantfirl <notifications@github.com> wrote:

> *@grantfirl* commented on this pull request.
> ------------------------------
>
> In physics/GFS_DCNV_generic.meta
> <https://github.com/NOAA-GSD/ccpp-physics/pull/5#discussion_r370329893>:
>
> > @@ -25,6 +25,14 @@
>    type = logical
>    intent = in
>    optional = F
> +[qdiag3d]
> +  standard_name = flag_tracer_diagnostics_3D
>
> Isn't it just specific humidity tendencies instead of all tracers?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-GSD/ccpp-physics/pull/5?email_source=notifications&email_token=AJMW4SL7PLBNPN3LQLARMNDQ7H2PLA5CNFSM4KCPQQJKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCS3ZIIY#pullrequestreview-347575331>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJMW4SLYTJ3R5RCG6MRF7G3Q7H2PLANCNFSM4KCPQQJA>
> .
>

SamuelTrahanNOAA commented:
Generic refers to the GFS generic interstitial "schemes."

On Thu, 23 Jan 2020 at 15:14, grantfirl <notifications@github.com> wrote:

> *@grantfirl* commented on this pull request.
> ------------------------------
>
> In physics/GFS_GWD_generic.meta
> <https://github.com/NOAA-GSD/ccpp-physics/pull/5#discussion_r370331972>:
>
> > @@ -145,6 +177,13 @@
>    kind = kind_phys
>    intent = in
>    optional = F
> +[gwd_generic_tend]
> +  standard_name = true_if_GFS_GWD_generic_should_calculate_tendencies
>
> Agree: flag_for_generic_gwd_tendencies is better, although "generic" is a
> bit vague in this context. You mean generic as opposed to tendencies that
> Valery put in for the CIRES UGWP scheme?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-GSD/ccpp-physics/pull/5?email_source=notifications&email_token=AJMW4SMD2ILYI75NF4WRMJ3Q7H3CHA5CNFSM4KCPQQJKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCS3Z5PA#discussion_r370331972>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJMW4SJOEPPAWCDW62W4GVTQ7H3CHANCNFSM4KCPQQJA>
> .
>

SamuelTrahanNOAA commented:
Same here.  When I found code I didn't understand, I tried to leave the old
behavior as closely as possible.

On Thu, 23 Jan 2020 at 15:20, grantfirl <notifications@github.com> wrote:

> *@grantfirl* commented on this pull request.
> ------------------------------
>
> In physics/GFS_MP_generic.F90
> <https://github.com/NOAA-GSD/ccpp-physics/pull/5#discussion_r370334751>:
>
> >            enddo
>          enddo
> -        do n=ntcw,ntcw+nncl-1
> -          save_q(1:im,:,n) = gq0(1:im,:,n)
> -        enddo
> +        if(qdiag3d) then
> +           do k=1,levs
> +              do i=1,im
> +                 ! Here, gq0(...,1) is used instead of gq0_water_vapor
> +                 ! to be consistent with the GFS_MP_generic_post_run
> +                 ! code.
> +                 save_qv(i,k) = gq0(i,k,1)
> +              enddo
> +           enddo
> +        endif
> +        if(do_aw) then
>
> I'm curious why this special treatment for do_aw...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-GSD/ccpp-physics/pull/5?email_source=notifications&email_token=AJMW4SIXOV6YMKW3C7LFKP3Q7H32HA5CNFSM4KCPQQJKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCS32ZKY#pullrequestreview-347581611>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJMW4SP6TZK7K4PRBZMHQZDQ7H32HANCNFSM4KCPQQJA>
> .
>

SamuelTrahanNOAA commented:
Passing in both gq0 and gq0_water_vapor will break assumptions in the
compiler's optimizer.  The resulting executable may not work.

On Thu, 23 Jan 2020 at 15:23, grantfirl <notifications@github.com> wrote:

> *@grantfirl* commented on this pull request.
> ------------------------------
>
> In physics/GFS_MP_generic.F90
> <https://github.com/NOAA-GSD/ccpp-physics/pull/5#discussion_r370335851>:
>
> >              enddo
>            enddo
> +          if (qdiag3d) then
> +             do k=1,levs
> +                do i=1,im
> +                   dq3dt(i,k) = dq3dt(i,k) + (gq0(i,k,1)-save_qv(i,k)) * frain
>
> similar comment to above. I'd prefer to see just the water vapor slice
> passed in.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-GSD/ccpp-physics/pull/5?email_source=notifications&email_token=AJMW4SIPMMSCTLTTZ7YCV23Q7H4EBA5CNFSM4KCPQQJKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCS33EMY#pullrequestreview-347583027>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJMW4SNA2DZN2JHLJNSN5MLQ7H4EBANCNFSM4KCPQQJA>
> .
>

SamuelTrahanNOAA commented:
The pbl_generic_tend flag is missing.

On Thu, 23 Jan 2020 at 15:35, grantfirl <notifications@github.com> wrote:

> *@grantfirl* commented on this pull request.
> ------------------------------
>
> In physics/moninedmf.f
> <https://github.com/NOAA-GSD/ccpp-physics/pull/5#discussion_r370340986>:
>
> > @@ -1037,6 +1041,17 @@ subroutine hedmf_run (ix,im,km,ntrac,ntcw,dv,du,tau,rtg,          &
>              rtg(i,k,1) = rtg(i,k,1)+qtend
>              dtsfc(i)   = dtsfc(i)+cont*del(i,k)*ttend
>              dqsfc(i)   = dqsfc(i)+conq*del(i,k)*qtend
> +            if(lssav .and. ldiag3d) then
>
> I'm a little confused why tendencies specific to moninedmf are needed.
> Shouldn't the ones calculated in GFS_PBL_generic suffice? If not, would you
> want to add ".and. .not. pbl_generic_tend"?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-GSD/ccpp-physics/pull/5?email_source=notifications&email_token=AJMW4SPLWAQ5W7BRPIH4OTTQ7H5R7A5CNFSM4KCPQQJKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCS34ZAQ#pullrequestreview-347589762>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJMW4SPU37L7MMWT2X4RPSDQ7H5R7ANCNFSM4KCPQQJA>
> .
>

SamuelTrahanNOAA commented:
Some schemes output their tendencies instead of modifying the state.  In
that situation, the diagnostic tendencies are calculated from the scheme's
output tendencies.  That is the ideal solution from a framework standpoint.

You cannot calculate the tendencies from state variables in the generic
interstitials because the generic interstitials don't know which tendencies
are provided by the scheme.  For example, you may have a PBL scheme that
also generates OGW tendencies.  In that case the interstitial has no idea
how to split the two.

Most importantly, schemes will want to split the generic tendencies (like
"convective gravity wave drag") into more logical components.  That is not
something that can be done in a generic interstitial because the generic
interstitial has no idea what the internal processes are within the scheme.

In other words, there are only two options:

1. A scheme outputs its tendency for that timestep instead of modifying the
state.  The interstitial accumulates those tendencies to make diagnostic
accumulated tendencies.

2. A scheme modifies the state directly, and also outputs the diagnostic
accumulated tendencies.

The current approach of modifying the state, and having the interstitial do
the math, does not work (except perhaps for microphysics).

Sincerely,
Sam Trahan

On Thu, 23 Jan 2020 at 15:55, grantfirl <notifications@github.com> wrote:

> I'm curious what the strategy for tendencies in general is. Process-split
> schemes like surface layer, PBL, and GWD must output a tendency while
> time-split schemes like the others don't necessarily have to except for
> diagnostic purposes, although they are certainly good diagnostics to have.
> My opinion (not that anyone asked for it) would be to basically treat all
> CCPP schemes like process-split schemes and to calculate tendencies by
> default automatically through the CCPP framework and to not allow schemes
> to modify the state internally. For time-split schemes, those tendencies
> could then simply be applied automatically after they're called too.
>
> With the exception of CIRES UGWP, we've been putting all diagnostic
> tendencies in the "scheme-generic" interstitials because presumably we'd
> like to have tendencies calculated in the same way for any scheme of a
> given type. IMO, relying on all schemes to calculate their own tendencies
> internally means a lot of repeated code and potential confusion...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-GSD/ccpp-physics/pull/5?email_source=notifications&email_token=AJMW4SM6AANLPPLKNK7BITLQ7H72VA5CNFSM4KCPQQJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJY2FSA#issuecomment-577872584>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJMW4SMZONOE572OLOLLWZTQ7H72VANCNFSM4KCPQQJA>
> .
>

SamuelTrahanNOAA reviewed (commented):
> I'm curious what the strategy for tendencies in general is. Process-split schemes like surface layer, PBL, and GWD must output a tendency while time-split schemes like the others don't necessarily have to except for diagnostic purposes, although they are certainly good diagnostics to have. My opinion (not that anyone asked for it) would be to basically treat all CCPP schemes like process-split schemes and to calculate tendencies by default automatically through the CCPP framework and to not allow schemes to modify the state internally. For time-split schemes, those tendencies could then simply be applied automatically after they're called too.
> 
> With the exception of CIRES UGWP, we've been putting all diagnostic tendencies in the "scheme-generic" interstitials because presumably we'd like to have tendencies calculated in the same way for any scheme of a given type. IMO, relying on all schemes to calculate their own tendencies internally means a lot of repeated code and potential confusion...

Some schemes output their tendencies instead of modifying the state.  In that situation, the diagnostic tendencies are calculated from the scheme's output tendencies.  That is the ideal solution from a framework standpoint.

You cannot calculate the tendencies from state variables in the generic interstitials because the generic interstitials don't know which tendencies are provided by the scheme.  For example, you may have a PBL scheme that also generates OGW tendencies.  In that case the interstitial has no idea how to split the two.

Most importantly, schemes will want to split the generic tendencies (like "convective gravity wave drag") into more logical components.  That is not something that can be done in a generic interstitial because the generic interstitial has no idea what the internal processes are within the scheme.

In other words, there are only two options:

1. A scheme outputs its tendency for that timestep instead of modifying the state.  The interstitial accumulates those tendencies to make diagnostic accumulated tendencies.

2. A scheme modifies the state directly, and also outputs the diagnostic accumulated tendencies.

The current approach of modifying the state, and having the interstitial do the math, does not work (except perhaps for microphysics).

SamuelTrahanNOAA committed:
correct issues in PR: put "flag_for_" at the beginning of flag names, remove dimensions of variables that may not be allocated, and add the missing do3dt_PBL variable

SamuelTrahanNOAA committed:
sync with gsd/develop

SamuelTrahanNOAA committed:
sync with gsd/develop again

SamuelTrahanNOAA committed:
Merge branch 'gsd/develop' into gsd/develop

SamuelTrahanNOAA commented:
This pull request will be replaced.
